# Operaciones entre arreglos.

Uno de los grandes aportes de NumPy al ecosistema de Python es su capacidad para realizar operaciones vectorizadas, es decir, aplicar cálculos directamente sobre arreglos completos, sin necesidad de iterar elemento por elemento. Esto no solo hace que el código sea más limpio y fácil de leer, sino que también aumenta enormemente la eficiencia computacional.

In [3]:
import numpy as np

## Operaciones elementales entre arreglos

vectores de igual tamaño:

In [26]:
a = np.array([10, 20, 30])
b = np.array([1, 2, 3])

print(a + b)
print(a * b)
print(a - b)
print(a / b)

[11 22 33]
[10 40 90]
[ 9 18 27]
[10. 10. 10.]


In [15]:
print(a.ndim) # una dimensión
print(b.ndim)
print(a.shape) # i3 elementos en una dimensión
print(b.shape)

1
1
(3,)
(3,)


Podemos realizar operaciones aritméticas directamente:

Estas operaciones se aplican elemento por elemento, y son mucho más rápidas que un bucle tradicional en Python.

## ¿Qué pasa si los tamaños no coinciden?

NumPy intentará aplicar una técnica llamada broadcasting. Veamos

In [27]:
a = np.array([10, 20, 30])
b = np.array([1]) # arreglo con un solo valor

print(a + b)
print(a * b)
print(a - b)
print(a / b)
# Resultado: [11 21 31]

[11 21 31]
[10 20 30]
[ 9 19 29]
[10. 20. 30.]


Aquí, b se "expande" automáticamente para coincidir con la forma de a.
Este comportamiento es útil, pero debemos tener cuidado cuando los
tamaños no son compatibles, ya que puede generar errores:

In [17]:
a = np.array([10, 20, 30])
b = np.array([1, 2])

#print(a + b) # Error: operands could not be broadcast together


## Operaciones entre matrices

Las mismas operaciones se aplican entre matrices del mismo tamaño:

In [29]:
m1 = np.array([[1, 2], [3, 4]])
m2 = np.array([[10, 20], [30, 40]])
print(m1 + m2)# Resultado:# [[11 22]# [33 44]]
print(m1* m2)
print(m1 - m2)
print(m1 / m2)

[[11 22]
 [33 44]]
[[ 10  40]
 [ 90 160]]
[[ -9 -18]
 [-27 -36]]
[[0.1 0.1]
 [0.1 0.1]]


También pueden realizarse combinaciones entre una matriz y un vector, si las dimensiones son compatibles (broadcasting):

In [33]:
# vector es de 2 elementos y la matriz es de 2*2
vector = np.array([1, 2]) # broadcasting: equivalente a np.array([[1,2],[1,2]])
print(vector)
print(m1)
print(f'\n', m1 * vector)

print(f'\n', m1 + vector)

arreglo = np.array([[1,2],[1,2]]) # esto es para comprobar que es igual que el vector
print(f'\n', m1 + arreglo)
# Resultado:
# [[2 4]
# [4 6]]

[1 2]
[[1 2]
 [3 4]]

 [[1 4]
 [3 8]]

 [[2 4]
 [4 6]]

 [[2 4]
 [4 6]]


## Operaciones lógicas y comparaciones

También puedes aplicar comparaciones directamente entre arreglos:

In [35]:
a = np.array([10, 20, 30])
b = np.array([10, 15, 30])
print(a == b) # [ True False True]
print(a > b) # [False True False]
print(a != b)

[ True False  True]
[False  True False]
[False  True False]


Esto devuelve un arreglo booleano, que puede ser usado para
filtrar datos.

In [39]:
# con np.where muestra el indice en dnde se cumple la condición 
np.where(a == b)

(array([0, 2]),)


In [40]:
np.where(a != b)

(array([1]),)

## Operaciones combinadas y funciones

Podemos encadenar operaciones con funciones adicionales:

In [44]:
print(a)
print(b)

[10 20 30]
[10 15 30]


In [43]:
resultado = (a + b) / 2
print(resultado)

[10.  17.5 30. ]


O utilizar funciones de agregación de NumPy:

In [46]:
# estas f(x) de agregación aplicarlos sobre vectores que esten represntando una misma varible o caracteristica
# ej: edades, humedad , temperaturas, etc. 
print(np.sum(a))# Suma total del vector a (10 + 20 +30) = 60
print(np.mean(a))# Promedio del vector (10 + 20 +30) = 60 / 3 = 20
print(np.max(a))# Valor máximo del vector , en este caso 30
print(np.std(a))# Desviación estándar del vector

60
20.0
30
8.16496580927726


## ¿Qué pasa con los tipos de datos?

NumPy realiza promoción de tipos automáticamente cuando es necesario:

In [49]:
a = np.array([1, 2, 3])
# int
b = np.array([0.5, 1.5, 2.5])
# float
resultado = a + b
print(resultado)
# Resultado: [1.5 3.5 5.5] → float

[1.5 3.5 5.5]


In [51]:
resultado.dtype

dtype('float64')

Esto asegura que los resultados mantengan la precisión
adecuada.

## Operaciones con escalares.

¿Cómo puedes transformar todos los valores de un arreglo aplicando
un mismo número?

En ciencia de datos es muy común aplicar operaciones uniformes a todos los elementos de una estructura. Esto puede incluir:

• Escalar todos los valores por un coeficiente.

• Sumar o restar un número fijo.

• Calcular proporciones o porcentajes.

• Aplicar constantes de normalización.

NumPy permite realizar estas tareas de forma directa gracias a la vectorización. Al operar un arreglo con un escalar, NumPy aplica automáticamente la operación a cada elemento del arreglo sin necesidad de bucles.

### Suma y resta con escalares

Cada valor del arreglo es modificado con la operación indicada.

In [55]:
import numpy as np

datos = np.array([10, 20, 30])

In [56]:
# 5 y 3 es el escalar
print(datos + 5) # [15 25 35]
print(datos - 3) # [ 7 17 27]

[15 25 35]
[ 7 17 27]


### Multiplicación y división con escalares

In [57]:
print(datos * 2) # el escalar es el 2 por lo que duplica el vector
# [20 40 60]
print(datos / 10) # e escalar es 10 por lo que divide en 10 cada elemento
# [1. 2. 3.]

[20 40 60]
[1. 2. 3.]


Este enfoque es ideal para convertir unidades, calcular proporciones o ajustar magnitudes.

### Operaciones de potenciación y módulo

In [58]:
print(datos ** 2) # potencia (elevado a ...) El escalar es el 2 y cada uno de los elementos se eleva al cuadrado.
# [100 400 900] → cuadrado
print(datos % 7) # el resto de la división.
# [3 6 2] → módulo 7

[100 400 900]
[3 6 2]


Estas operaciones también se aplican elemento por elemento.

### Aplicaciones frecuentes

- Ajuste de precios

In [60]:
precios = np.array([100, 200, 300])
nuevos = precios * 1.05 # Aumento de 5%

print(nuevos)

[105. 210. 315.]


✅ Ejemplo: aumento de 5%

5% en decimal es:

5 / 100 = 0.05


Entonces el nuevo valor es:

valor original + 5% del valor original


En fórmula:

valor * (1 + 0.05)

Que es lo mismo que:

valor * 1.05

✅ Por eso se usa:
precios * 1.05


porque:

1 → es el 100% del precio original

0.05 → es el 5% de aumento

👉 1 + 0.05 = 1.05

- Conversión de temperaturas

In [62]:
celsius = np.array([0, 10, 20, 30])
fahrenheit = celsius * 9 / 5 + 32

print(fahrenheit)

[32. 50. 68. 86.]


- Normalización de datos

In [63]:
datos = np.array([5, 10, 15])
normalizados = datos / np.max(datos) # Escala entre 0 y 1

print(normalizados)

[0.33333333 0.66666667 1.        ]


✅ Paso 2: ¿qué hace np.max(datos)?

Busca el valor más grande: en este caso 15

✅ Paso 3: qué hace esta línea
normalizados = datos / np.max(datos)

Es exactamente lo mismo que:

normalizados = datos / 15

✅ Paso 4: división elemento a elemento

NumPy divide cada elemento del arreglo por 15:

- 5  / 15 = 0.33333333

- 10 / 15 = 0.66666667

- 15 / 15 = 1

🧠 ¿Por qué esto se llama normalización?

Porque estás llevando los datos a una escala común:

👉 el valor más grande queda en 1
👉 el resto queda entre 0 y 1

🎯 En palabras simples

Esta forma de normalizar significa:

cada valor se expresa como una fracción del valor máximo.

🔎 Interpretación rápida

Tu arreglo original era:

[5, 10, 15]


Y ahora significa:

- 5 es el 33% del máximo

- 10 es el 66% del máximo

- 15 es el 100% del máximo

## Errores comunes

### División entera sin esperarlo

In [64]:
a = np.array([5, 10])
print(a / 2)
# Resultado esperado: [2.5 5.0]


[2.5 5. ]


Pero si a fuera de tipo int32 y usas división entera //, perderás decimales:

In [65]:
print(a // 2)
# Resultado: [2 5] → división truncada

[2 5]


### Tipo de dato inesperado

In [66]:
a = np.array([True, False])
print(a * 10)
# Resultado: [10 0] → True = 1, False = 0

[10  0]


NumPy interpreta los valores booleanos como enteros.